src/TimeTrace/ExecutionHierarchy.cpp

```diff
@@ -0,0 +1,342 @@
+#include "ExecutionHierarchy.h"
+
+#include <assert.h>
+#include <string>
+
+using namespace Microsoft::Cpp::BuildInsights;
+using namespace Activities;
+using namespace SimpleEvents;
+
+using namespace vcperf;
+
+namespace
+{
+    std::string ToString(const std::wstring& wstring)
+    {
+        assert(!wstring.empty());
+
+        const UINT codePage = CP_UTF8;
+        int requiredSize = WideCharToMultiByte(codePage, 0, wstring.c_str(), static_cast<int>(wstring.size()),
+                                               NULL, 0, NULL, NULL);
+        std::string convertedString = std::string(requiredSize, '\0');
+        WideCharToMultiByte(codePage, 0, wstring.c_str(), static_cast<int>(wstring.size()),
+                            &convertedString[0], requiredSize, NULL, NULL);
+
+        return convertedString;
+    }
+
+    long long ConvertTickPrecision(long long ticks, long long fromFreq, long long toFreq)
+    {
+        if (fromFreq <= 0) {
+            return 0;
+        }
+
+        long long p1 = (ticks / fromFreq) * toFreq;
+        long long p2 = (ticks % fromFreq) * toFreq / fromFreq;
+
+        return p1 + p2;
+    }
+
+    std::chrono::nanoseconds ConvertTime(long long ticks, long long frequency)
+    {
+        return std::chrono::nanoseconds{ ConvertTickPrecision(ticks, frequency, std::chrono::nanoseconds::period::den) };
+    }
+
+}  // anonymous namespace
+
+bool ExecutionHierarchy::Entry::OverlapsWith(const Entry* other) const
+{
+    return StartTimestamp < other->StopTimestamp &&
+           other->StartTimestamp < StopTimestamp;
+}
+
+ExecutionHierarchy::ExecutionHierarchy(const Filter& filter) :
+    entries_{},
+    roots_{},
+    filter_{filter},
+    symbolNames_{},
+    unresolvedTemplateInstantiationsPerSymbol_{}
+{
+}
+
+AnalysisControl ExecutionHierarchy::OnStartActivity(const EventStack& eventStack)
+{
+    if (   MatchEventStackInMemberFunction(eventStack, this, &ExecutionHierarchy::OnNestedActivity)
+        || MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnRootActivity))
+    {}
+
+    if (   MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnInvocation)
+        || MatchEventInMemberFunction(eventStack.Back(), this, &ExecutionHierarchy::OnFrontEndFile))
```

Comment:
The display of Thread activities in the JSON file is a bit ambiguous. Sometimes a thread can be either from CodeGeneration or BottomUp, TopDown, etc, but right now only "Thread" is displayed. When generating the entry for Thread, would it be possible to concatenate the name of the parent activity to know which kind of thread it is? E.g. CodeGenerationThread, etc. You can probably do this with an `OnThread(const Activity& parent, const Thread& thread)` function.

Reply:
Sure, no problem with that 😉